lib/srv/db/sqlserver/connect_test.go

```diff
@@ -116,3 +125,257 @@ func TestConnectorSelection(t *testing.T) {
+		DBAuth:                &mockDBAuth{},
+		AuthClient:            &mockAuth{},
+		kinitCommandGenerator: &staticCache{t: t, pass: true},
+		caFunc: func(ctx context.Context, clusterName string) ([]byte, error) {
+			return []byte(mockCA), nil
+		},
+	}
+
+	krbConfPath := filepath.Join(dir, "krb5.conf")
+	err := os.WriteFile(krbConfPath, []byte(krb5Conf), 0664)
+	require.NoError(t, err)
+
+	for _, tt := range []struct {
+		desc         string
+		databaseSpec types.DatabaseSpecV3
+		errAssertion require.ErrorAssertionFunc
+	}{
+		{
+			desc: "AD-x509-Loads_and_fails_with_expired_cache",
+			databaseSpec: types.DatabaseSpecV3{
+				Protocol: defaults.ProtocolSQLServer,
+				URI:      "sqlserver:1443",
+				AD: types.AD{
+					LDAPCert:    mockCA,
+					KDCHostName: "kdc.example.com",
+					Krb5File:    krbConfPath,
+				},
+			},
+			// When using a non-Azure database, the connector should attempt to get a kinit client
+			errAssertion: func(t require.TestingT, err error, _ ...interface{}) {
+				require.Error(t, err)
+				// we can't get a new TGT without an actual kerberos implementation, so we are relying on the existing
+				// credentials cache being expired
+				require.ErrorContains(t, err, "cannot login, no user credentials available and no valid existing session")
+			},
+		},
+		{
+			desc: "AD-x509-Fails_to_load_with_bad_config",
+			databaseSpec: types.DatabaseSpecV3{
+				Protocol: defaults.ProtocolSQLServer,
+				URI:      "sqlserver:1443",
+				AD:       types.AD{},
+			},
+			// When using a non-Azure database, the connector should attempt to get a kinit client
+			errAssertion: func(t require.TestingT, err error, _ ...interface{}) {
+				require.Error(t, err)
+				// we can't get a new TGT without an actual kerberos implementation, so we are relying on the existing
+				// credentials cache being expired
+				require.ErrorIs(t, err, errBadKerberosConfig)
+			},
+		},
+		{
+			desc: "AD-x509-Fails_with_invalid_certificate",
+			databaseSpec: types.DatabaseSpecV3{
+				Protocol: defaults.ProtocolSQLServer,
+				URI:      "sqlserver:1443",
+				AD: types.AD{
+					LDAPCert:    "BEGIN CERTIFICATE",
+					KDCHostName: "kdc.example.com",
+					Krb5File:    krbConfPath,
+				},
+			},
+			// When using a non-Azure database, the connector should attempt to get a kinit client
+			errAssertion: func(t require.TestingT, err error, _ ...interface{}) {
+				require.Error(t, err)
+				// we can't get a new TGT without an actual kerberos implementation, so we are relying on the existing
+				// credentials cache being expired
+				require.ErrorIs(t, err, errBadCertificate)
+			},
+		},
+	} {
+		t.Run(tt.desc, func(t *testing.T) {
+			database, err := types.NewDatabaseV3(types.Metadata{
+				Name: uuid.NewString(),
+			}, tt.databaseSpec)
+			require.NoError(t, err)
+
+			databaseUser := "alice"
+			databaseName := database.GetName()
+			connector.DataDir = dir
+
+			connectorCtx, cancel := context.WithCancel(ctx)
+			cancel()
```

Comment:
Probably worth including a short comment that explains why we are creating a cancelled context here.